### ISSUE TITLE
Using XbimGeometry with mono

### ISSUE THREAD

**Issue by VegardAaberge** [xBimTeam/XbimGeometry#56]
On Windows I am able to run XbimGeometry just fine, but when I try to run it on Visual Studio for Mac then I am only able to run XbimEssentials, not XbimGeometry. 

When I run it on Visual Studio for Mac then this line fails.
`private readonly XbimGeometryEngine _xbimGeometryCreator = new XbimGeometryEngine()`

with the error 
> The Visual C++ runtime 'VC12' (VS2013) could not be located. This is required for XBim Geometry generation. Please install this from https://www.microsoft.com/en-us/download/details.aspx?id=40784 
> 
> To suppress this exception define an AppSetting with key of 'SuppressVCRuntimeWarning' and value of 'true' in app.config

I cannot install an exe file on mac, and supressing that exception just lead to another exception. Is there any way I can run XbimGeometry with mono? Or do I need to replace the XbimGeometry functions that I used with XbimEssentials functions?

**Comment by CBenghi** [comment]
As far as I know we have a long term roadmap to port the solution to Mono, however this depends on Mono's essential features; the geometry engine relies on is a mixed mode c++ project that is currently not supported by Mono.
You can read more about this at http://www.mono-project.com/docs/about-mono/languages/cplusplus/
For the time being your best option would be to establish a windows web service to perform the conversion online, please note that another essential bit of the platform (Esent) is also (as far as I know) not compatible with Mac, so you might want to extend Essential to use a different persistency layer (Sqlite maybe?).
In short, it's going to be a while before xbim works on a mac.

**Comment by VegardAaberge** [comment]
Thank you for the clear answer. 

Shamefully we cannot use a windows service, so I have to use mono. Luckily Xbim Essentials works well with mono, because they have developed an in-memory model that is platform independent.
https://github.com/xBimTeam/XbimEssentials/issues/32

**Comment by toddams** [comment]
I'm getting the same exception. So what should I do in order to get in working with mono, as you are saying it works well? Thanks

**Comment by VegardAaberge** [comment]
hey toddams. 

What I did was to stop using XbimGeometry, and replaced the methods I used from XbimGeometry . You can look at the source code in XbimGeometry, and imeplement those methods yourself.

Also, remember to add some parameter to IfcStore.Open, because else it will crash when file becomes too big.
`using (var model = IfcStore.Open(ifcFile, null, -1))`

**Comment by toddams** [comment]
Our code heavily relies on XbimRect3D, dunno if I can find alternative at xBimEssentials. Anyway, thank you very much for such a quick response ;)

**Comment by toddams** [comment]
I have the only reference to Geometry library - this is Xbim3DModelContext.cs class.

````
var context = new Xbim3DModelContext(model);
context.CreateContext();
````

This is the only line that uses it. To be honest, I don't know what it does) This project is managed by another team, my only responsibility is to port it to Mono. Maybe you can give me a hint and tell how can I get rid of that line and not ruin the entire project? 

btw, thank you very much for such a quick response :)

**Comment by VegardAaberge** [comment]
Is context being used again? You need to remove context and get data directly from the model like this
`var site = model.Instances.OfType<IfcSpatialStructureElement>().OfType<IfcSite>().First();`

Just try to remove XbimGeometry and those lines. Then you will get some errors, just disable the functionality that use xBim Geometry and reenable them step by step. If it uses a lot of XbimGeometry functionality, then I would follow CBenghi advice.

**Comment by toddams** [comment]
Actually no, context is not used anymore anywhere. Just these 2 lines of code :)

**Comment by CBenghi** [comment]
Perhaps the context is not used any more, but its actions are meaningful.
The generation of the context also persists the meshed model to the database, so without those two lines you would not be able to display graphics of the model at all.
If this is ok then drop geometry all-together and use essentials alone. 
Otherwise my advice above stands.

**Comment by toddams** [comment]
````
private static byte[] CreateWexBIMFile(IfcStore model)
{
    //var context = new Xbim3DModelContext(model);
    //context.CreateContext();

     MemoryStream objMemoryStream = new MemoryStream();
     using (var wexBimBinaryWriter = new BinaryWriter(objMemoryStream))
     {
          model.SaveAsWexBim(wexBimBinaryWriter);
          wexBimBinaryWriter.Close();
     }
    return objMemoryStream.ToArray();
}
````
Yea, if I commend this, model is generated, I can see the components of it, but on WebGL viewer it's not displaying it. 

![](https://i.gyazo.com/2d4e2e731ae37f90c2030f140b44502d.png)

Seems like there's really no way how to workaround this :(

Anyway, thank you for your help)

**Comment by martin1cerny** [comment]
No, all ways around involve heavy refactoring and coding.
